### PR TITLE
feat(github): Add autodiscovery support for Github App

### PR DESCRIPTION
**Pull request by DiabolicaTrix**
## Changes:
The endpoint used to get a list of the accessible repositories is changed from `user/repos` to `installation/repositories` if a Github App is used and the response from that new endpoint is converted to be handled properly.

## Context:
The autodiscovery feature uses the `user/repos` endpoint to get a list of repositories, but that endpoint does not work with a Github App. In that case, the endpoint needs to be changed to `installation/repositories`.

Unfortunately, the response from these 2 endpoints is not exactly the same.
`user/repos` directly gives a list of repositories while the list from `installation/repositories` is in the `repositories` field.
```
{
  ...
  "repositories": [...]
}
```
Apart from that small difference, the response is exactly the same and works without any other modifications.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Closes #13467

### CODE REVIEW

**Comment by rarkins** [comment]
Please create a feature request issue which this PR closes, and mention your requirements and implementation approach there.

File: lib/util/http/github.ts
Comment (rarkins) [inline lib/util/http/github.ts:211]: This is not appropriate for this file as it should not be field-aware

Comment (alexymantha) [reply lib/util/http/github.ts]: Thanks for the feedback. You are right, and I also realized that it was not necessary because I can use the paginationField to achieve the same result.

Comment (DiabolicaTrix) [reply lib/util/http/github.ts]: Thanks for the feedback. You are right, and I also realized that it was not necessary because I can use the paginationField to achieve the same result.

**Comment by DiabolicaTrix** [comment]
Blocked by #13474 paginationField stopped working after PR #13416 was merged

**Comment by zharinov** [comment]
Suggested edit:
```diff
diff --git a/lib/platform/github/index.ts b/lib/platform/github/index.ts
index 03b0c3398..e0195f2a4 100644
--- a/lib/platform/github/index.ts
+++ b/lib/platform/github/index.ts
@@ -152,15 +152,21 @@ export async function getRepos(): Promise<string[]> {
   });
   try {
     const isGHApp = opts.token?.startsWith('x-access-token:');
-    const endpoint = isGHApp ? 'installation/repositories' : 'user/repos';
-    const res = await githubApi.getJson<{ full_name: string }[]>(
-      `${endpoint}?per_page=100`,
-      { paginationField: 'repositories', paginate: 'all' }
-    );
     if (isGHApp) {
-      res.body = res.body.repositories;
+      const res = await githubApi.getJson<{
+        repositories: { full_name: string }[];
+      }>(`installation/repositories?per_page=100`, {
+        paginationField: 'repositories',
+        paginate: 'all',
+      });
+      return res.body.repositories.map((repo) => repo.full_name);
+    } else {
+      const res = await githubApi.getJson<{ full_name: string }[]>(
+        `user/repos?per_page=100`,
+        { paginate: 'all' }
+      );
+      return res.body.map((repo) => repo.full_name);
     }
-    return res.body.map((repo) => repo.full_name);
   } catch (err) /* istanbul ignore next */ {
     logger.error({ err }, `GitHub getRepos error`);
     throw err;

```

File: lib/util/http/github.spec.ts
Comment (zharinov) [inline lib/util/http/github.spec.ts:156]: Test should reside here: https://github.com/renovatebot/renovate/blob/7c4a71b679453af716fdb4fb96425dca1e712188/lib/platform/github/index.spec.ts#L161

**Review by viceice — CHANGES_REQUESTED** [review]
See comments from @zharinov

**Comment by DiabolicaTrix** [comment]
I made the requested changes so the PR itself is ready for review but will not work until #13474 is fixed so I think I should leave it as a draft for the moment

**Comment by DiabolicaTrix** [comment]
#13506 fixed it, so it is not blocked anymore and ready for review. Thanks @zharinov!

File: lib/platform/github/index.spec.ts
Comment (viceice) [inline lib/platform/github/index.spec.ts:199]:
```suggestion
```

no longer needed.

File: lib/platform/github/index.spec.ts
Comment (viceice) [inline lib/platform/github/index.spec.ts:198]:
```suggestion
      expect(repos).toStricEqual([
  "a/b",
  "c/d",
]);
```

needs prettify and snapshot update

File: lib/platform/github/index.ts
Comment (viceice) [inline lib/platform/github/index.ts:154]: I think this detection should happen in `initPlatform` and stored to `platformConfig`. So we can reuse later where required, like `gheVersion`.

https://github.com/renovatebot/renovate/blob/a3635b7d9ebdcfb7c122cc6b8738cd586250b0ea/lib/platform/github/index.ts#L452-L455

@rarkins WDYT?

Comment (DiabolicaTrix) [reply lib/platform/github/index.ts]: I agree that's a better place to do the detection. I moved it to initPlatform in the latest commits, let's see what @rarkins thinks

File: lib/platform/github/index.ts
Comment (viceice) [inline lib/platform/github/index.ts:112]: @rarkins We should use this in `initRepo` in a separate pr, so the code there will be simplified.

**Review by viceice — APPROVED** [review]

File: lib/platform/github/index.ts
Comment (rarkins) [reply lib/platform/github/index.ts:112]: OK. I need to double check that these changes don't break the app or WSOP too

**Review by rarkins — APPROVED** [review]

**Pull request merged**
